micrometer-observation/src/main/java/io/micrometer/observation/Observation.java

```diff
@@ -367,7 +366,9 @@ default Observation lowCardinalityKeyValue(String key, String value) {
      * @return this
      */
     default Observation lowCardinalityKeyValues(KeyValues keyValues) {
-        keyValues.stream().forEach(this::lowCardinalityKeyValue);
```

Comment:
the Iterable.forEach default implementation makes a [for loop internally](https://github.com/openjdk/jdk/blob/a9a271179d2a7952154b7509a999b100cc98b13c/src/java.base/share/classes/java/lang/Iterable.java#L72-L77), so yeah it is generally equivalent. however using just a plain for loop is slightly more efficient since we skip that nullcheck and also `this::` lambda need not be created and invoked.

here is a test class:
````java
interface ThisLambdaTest {

	void lowCardinalityKeyValue(String v);

	default void lowCardinalityKeyValues(Iterable<String> keyValues) {
		keyValues.forEach(this::lowCardinalityKeyValue);
	}

	default void lowCardinalityKeyValues2(Iterable<String> keyValues) {
		for (String keyValue : keyValues) {
			lowCardinalityKeyValue(keyValue);
		}
	}
}
````

and here is the bytecode:
````
// Method descriptor #8 (Ljava/lang/Iterable;)V
  // Signature: (Ljava/lang/Iterable<Ljava/lang/String;>;)V
  // Stack: 2, Locals: 2
  public void lowCardinalityKeyValues(java.lang.Iterable keyValues);
     0  aload_1 [keyValues]
     1  aload_0 [this]
     2  invokedynamic 0 accept(com.triplelift.exchange.auction.ThisLambdaTest) : java.util.function.Consumer [12]
     7  invokeinterface java.lang.Iterable.forEach(java.util.function.Consumer) : void [16] [nargs: 2]
    12  return
      Line numbers:
        [pc: 0, line: 8]
        [pc: 12, line: 9]
      Local variable table:
        [pc: 0, pc: 13] local: this index: 0 type: com.triplelift.exchange.auction.ThisLambdaTest
        [pc: 0, pc: 13] local: keyValues index: 1 type: java.lang.Iterable
      Local variable type table:
        [pc: 0, pc: 13] local: keyValues index: 1 type: java.lang.Iterable<java.lang.String>
  
  // Method descriptor #8 (Ljava/lang/Iterable;)V
  // Signature: (Ljava/lang/Iterable<Ljava/lang/String;>;)V
  // Stack: 2, Locals: 4
  public void lowCardinalityKeyValues2(java.lang.Iterable keyValues);
     0  aload_1 [keyValues]
     1  invokeinterface java.lang.Iterable.iterator() : java.util.Iterator [31] [nargs: 1]
     6  astore_3
     7  goto 27
    10  aload_3
    11  invokeinterface java.util.Iterator.next() : java.lang.Object [35] [nargs: 1]
    16  checkcast java.lang.String [41]
    19  astore_2 [keyValue]
    20  aload_0 [this]
    21  aload_2 [keyValue]
    22  invokeinterface com.triplelift.exchange.auction.ThisLambdaTest.lowCardinalityKeyValue(java.lang.String) : void [43] [nargs: 2]
    27  aload_3
    28  invokeinterface java.util.Iterator.hasNext() : boolean [45] [nargs: 1]
    33  ifne 10
    36  return
      Line numbers:
        [pc: 0, line: 12]
        [pc: 20, line: 13]
        [pc: 27, line: 12]
        [pc: 36, line: 15]
      Local variable table:
        [pc: 0, pc: 37] local: this index: 0 type: com.triplelift.exchange.auction.ThisLambdaTest
        [pc: 0, pc: 37] local: keyValues index: 1 type: java.lang.Iterable
        [pc: 20, pc: 27] local: keyValue index: 2 type: java.lang.String
      Local variable type table:
        [pc: 0, pc: 37] local: keyValues index: 1 type: java.lang.Iterable<java.lang.String>
      Stack map table: number of frames 2
        [pc: 10, full, stack: {}, locals: {com.triplelift.exchange.auction.ThisLambdaTest, java.lang.Iterable, _, java.util.Iterator}]
        [pc: 27, same]

  Inner classes:
    [inner class info: #67 java/lang/invoke/MethodHandles$Lookup, outer class info: #69 java/lang/invoke/MethodHandles
     inner name: #71 Lookup, accessflags: 25 public static final]
Bootstrap methods:
  0 : # 61 invokestatic java/lang/invoke/LambdaMetafactory.metafactory:(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodHandle;Ljava/lang/invoke/MethodType;)Ljava/lang/invoke/CallSite;
	Method arguments:
		#63 (Ljava/lang/Object;)V
		#64 null
		#65 (Ljava/lang/String;)V
````

the lambda and invokedynamic is only present in the forEach version, whereas the iterator is the only allocation in the for version.